### PR TITLE
Add a Configure option to enable strict by default in perl programs.

### PR DESCRIPTION
**Pull request by toddr**
Do not set strict by default on `-e` or `-E`

This is a development only feature to allow us to clean up blead.

`./Configure -Dusedevel -des -Dusedefaultstrict`

### CODE REVIEW

File: pp_ctl.c [hunk 1]
```diff
@@ -3485,7 +3485,7 @@ S_doeval_compile(pTHX_ U8 gimme, CV* outside, U32 seq, HV *hh)
 
     SAVEHINTS();
     if (clear_hints) {
-	PL_hints = 0;
+	PL_hints = HINT_DEFAULT;
```
Comment (toddr) [inline pp_ctl.c]: @tonycoz I am 99% certain this is the correct thing to do here but admit to not being clear what the current circumstances are that we want to set it to 0?

**Comment by toddr** [comment]
@Tux needs Configure ❤️

File: perl.c [hunk 2]
```diff
@@ -2548,6 +2550,8 @@ S_parse_body(pTHX_ char **env, XSINIT_t xsinit)
 	filter_add(read_e_script, NULL);
 
     /* now parse the script */
+    if (minus_e == FALSE)
+        PL_hints |= HINT_DEFAULT; /* after init_main_stash ; need to be after init_predump_symbols */
```
Comment (toddr) [inline perl.c]: Currently perl programs rely on `PL_hints` being set to 0 by default. As a result there is no current code that initializes it. But we can't initialize it until we know if perl was run with `-e` or `-E`

**Review by xsawyerx — APPROVED** [review]
As it is for internal use, I think there isn't much bikeshedding to do on the configure option name (though I wouldn't object to other names, if there are good suggestions).

This ability strikes me as very useful for us to improve the state of our internal code, making sure more of our code and tests conform with `use strict`, as we recommend others use.

**Review by xsawyerx — CHANGES_REQUESTED** [review]
My approval is conditioned on this passing and being approved by at least two core developers. I suggest @tonycoz and @Tux.

**Comment by Leont** [comment]
A name starting with `d_` doesn't seem logical, it generally means "is something defined". It would be more consistent with existing options to use a name starting with `use` (e.g. `usethreads`, `usequadmath`, `usedevel`).

My intuition says that those default hint flags should be introspectable via a `${^HINT_DEFAULT}` that would be `0` if this feature isn't enabled.

**Comment by toddr** [comment]
> My intuition says that those default hint flags should be introspectable via a `${^HINT_DEFAULT}` that would be `0` if this feature isn't enabled.

variable renamed. 

Is this good enough in leu of a new global?

```
$>./perl -Ilib -MConfig -E'say "ok" if exists $Config{usedefaultstrict}'
ok
```

**Comment by Grinnz** [comment]
If we want to clean up blead to be strict safe, why not just add use strict to everywhere? No argument but this seems the overcomplex approach (in code style and clarity, as well as implementation).

**Comment by toddr** [comment]
> t add use strict to everywhere? No argument but this seems the overcomplex approach (in code style and clarity, as well as implementation).

Part of the research done elsewhere reveals that there are many corner cases that go unnoticed when you just put 'use strict' all over the place.

**Comment by Grinnz** [comment]
Can you elaborate? The only difference this would make from that approach is that things that haven't had explicit strict added yet would also break.

**Comment by toddr** [comment]
> Can you elaborate? The only difference this would make from that approach is that things that haven't had explicit strict added yet would also break.

This is a canary tool to help us better detect the less obvious places that do not have strict.

**Comment by Grinnz** [comment]
strict is lexical; thus you can detect all files that do not have strict with grep, or am I missing something? Things messing with `$^H` directly would be more complicated of course.

**Review by Tux — COMMENTED** [review]
I don't like the name, but that is bikeshedding. The patch looks sane, but if we want to "keep" this, it has to be turned into a meta-unit in order to get included the next time too.
So I won't object (other than to the name). Why not simply what it does `usedefaultstrict`

**Comment by toddr** [comment]
> So I won't object (other than to the name). Why not simply what it does `usedefaultstrict`

Because down the road, I was thinking about adding warnings at a minimum. So I need something more generic. I don't care about the name but `usedefaultstrict` kinda boxes me in for future changes.

**Comment by leonerd** [comment]
I'd suggest not bundling all the different bits into one option, but instead keep them as separate things. You might sometime discover that you want to do a canary run with warnings but not strict, because of one weird cornercase you want to skip over for the moment. Being able to separately `--usedefaultwarnings --nodefaultstrict` would be useful there.

**Comment by toddr** [comment]
> I'd suggest not bundling all the different bits into one option

Fair enough. Update coming. I'm going to split this into 2 commits with the first commit only being the changes required to enhance Configure. Hopefully that'll make it easier for people to read.

**Comment by toddr** [comment]
I've switched it to `./Configure -Dusedevel -des -Dusedefaultstrict` and updated the docs.

File: pp_ctl.c (hunk 1 above)
Comment (tonycoz) [reply pp_ctl.c]: They're set to 0 for require, so this change looks right.

**Review by tonycoz — APPROVED** [review]

**Comment by Leont** [comment]
> Is this good enough in leu of a new global?

Having the full hints flags allows for an overview of a wide variety of defaults, not just strictness but also feature bundles and utf8-ness.

That said, it's probably better as a separate PR anyway.

**Comment by toddr** [comment]
@Tux the configure option has been renamed. Are you ok with this now?

File: Configure [hunk 3]
```diff
@@ -24168,6 +24197,7 @@ d_eofnblk='$d_eofnblk'
 d_erf='$d_erf'
 d_erfc='$d_erfc'
 d_eunice='$d_eunice'
+usedefaultstrict='$usedefaultstrict'
```
Comment (Leont) [inline Configure:24200]: This should probably move down to keep the list alphabetically ordered?

Comment (toddr) [reply Configure:24200]: yep. we have a test for this but apparently it doesn't cover Configure.

Comment (toddr) [reply Configure:24200]: Actually on review of the vars, I see no sorting at all for these variables. I suspect it's going to get clobbered when @Tux adds his module so I'm not sure it's worth doing.

**Comment by Tux** [comment]
> @Tux the configure option has been renamed. Are you ok with this now?

Yes I am

File: Configure (hunk 3 above)
Comment (Tux) [reply Configure:24200]: It is not. Will be re-ordered later.

**Review by Tux — APPROVED** [review]

**Pull request merged**